drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,443 @@
+#endif /* DT_INST_NODE_HAS_PROP(0, reg_on_gpios) */
+
+	return 0;
+}
+
+int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface,
+			    whd_netif_funcs_t *netif_funcs, whd_buffer_funcs_t *buffer_if)
+{
+	int ret;
+	struct airoc_wifi_data *data = dev->data;
+	const struct airoc_wifi_config *config = dev->config;
+
+	whd_sdio_config_t whd_sdio_config = {
+		.sdio_1bit_mode = WHD_FALSE,
+		.high_speed_sdio_clock = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
+	whd_oob_config_t oob_config = {
+		.host_oob_pin = (void *)&config->wifi_host_wake_gpio,
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_sdio_config.oob_config = oob_config;
+#endif
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on retuens fail");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->sdhc_dev)) {
+		LOG_ERR("SDHC device is not ready");
+		return -ENODEV;
+	}
+
+	ret = sd_init(config->sdhc_dev, &data->card);
+	if (ret) {
+		return ret;
+	}
+
+	cy_rslt_t whd_ret = whd_init(&data->whd_drv, &init_config_default, &resource_ops, buffer_if,
+				     netif_funcs);
+	if (whd_ret == CY_RSLT_SUCCESS) {
+		whd_ret = whd_bus_sdio_attach(data->whd_drv, &whd_sdio_config,
+					      (whd_sdio_t)&data->card);
+
+		if (whd_ret == CY_RSLT_SUCCESS) {
+			whd_ret = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (whd_ret != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+			return -ENODEV;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Implement SDIO CMD52/53 wrappers
+ */
+whd_result_t whd_bus_sdio_cmd52(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, uint32_t address, uint8_t value,
+				sdio_response_needed_t response_expected, uint8_t *response)
+{
+	int ret;
+	struct sd_card *sd = whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+	struct sdhc_command cmd = {
+		.opcode = SDIO_RW_DIRECT,
+		.response_type = SD_RSP_TYPE_R5,
+		.timeout_ms = CONFIG_SD_CMD_TIMEOUT
+	};
+
+	arg.cmd52.function_number = (uint32_t)function;
+	arg.cmd52.register_address = (uint32_t)address;
+	arg.cmd52.rw_flag = (uint32_t)((direction == BUS_WRITE) ? 1 : 0);
+	arg.cmd52.write_data = value;
+	cmd.arg = arg.value;
+	WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd52);
+
+	ret = sdhc_request(sd->sdhc, &cmd, NULL);
+
+	WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(whd_driver->bus_priv, (ret != WHD_SUCCESS),
+						     cmd52_fail);
+	if (response != NULL) {
+		*response = cmd.response[0U] & SDIO_DIRECT_CMD_DATA_MASK;
+	}
+
+	/* Possibly device might not respond to this cmd. So, don't check return value here */
+	if ((ret != WHD_SUCCESS) && (address == SDIO_SLEEP_CSR)) {
+		return ret;
+	}
+
+	CHECK_RETURN(ret);
+	return WHD_SUCCESS;
+}
+
+whd_result_t whd_bus_sdio_cmd53(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
```

Comment:
Same question here- drivers should not be calling `sdhc_request` directly. CMD53 should be achieved with `sdio_write_fifo`, `sdio_read_fifo`, `sdio_read_addr`, and `sdio_write_addr`.

If these interfaces don't work for you, I'm open to modifying the API if needed. However, I don't want drivers to effectively reimplement parts of the SDIO stack within their own code

Reply:
Hi @danieldegrasse , sd subsystem api requires to enable sdio function (sdio_enable_func) before using rw functions.

Our WHD driver care about enabling function F0, F1, F2 in whd_bus_sdio_init. For sure we do not want to change this because of maintains reason.

For now if i try to call sdio_enable_func() for F1, F2 from airoc wifi driver in airoc_wifi_init_primary (before whd_wifi_on) i get stuck in sdio_read_cis() function of sd subsystem. I thinks it can because of accessing to F1, F2 before wifi fw is loaded.. Let me investigate/confirm this.

Reply:
Thank you for clarifying the issue a bit here. In Linux, SDIO CIS tuples are read during initialization: https://github.com/torvalds/linux/blob/master/drivers/mmc/core/sdio.c#L1276. I'm not familiar with the WHD driver architecture, but is AIROC supported by Linux? That may be somewhere to look. Another thing to check would be if the AIROC driver sets the `MMC_QUIRK_NONSTD_SDIO`, since then the CIS tuples would not be read: https://github.com/torvalds/linux/blob/master/drivers/mmc/core/sdio.c#L117

Reply:
Hi @danieldegrasse, thanks for Linux reference. I found the root cause of CIS reading problem. Added comments to SDIO/SD https://github.com/zephyrproject-rtos/zephyr/pull/56869. 

So with minor updates of sdio.c (https://github.com/npal-cy/zephyr-1/commit/50ad03a0cb9b1ec878ad6133014024cd99058ec8) , i could using sdio_read_byte/sdio_write_byte, sdio_read_addr/ sdio_write_addr for wrapping whd_bus_sdio_cmd52/whd_bus_sdio_cmd53.